server/base/src/main/java/org/apache/accumulo/server/util/HdfsTableDiskUsage.java

```diff
@@ -61,94 +60,16 @@
-  private Map<Integer,TableId> externalIds = new HashMap<>();
-  private Map<String,Integer[]> tableFiles = new HashMap<>();
-  private Map<String,Long> fileSizes = new HashMap<>();
-
-  void addTable(TableId tableId) {
-    if (internalIds.containsKey(tableId))
-      throw new IllegalArgumentException("Already added table " + tableId);
-
-    // Keep an internal counter for each table added
-    int iid = nextInternalId++;
-
-    // Store the table id to the internal id
-    internalIds.put(tableId, iid);
-    // Store the internal id to the table id
-    externalIds.put(iid, tableId);
-  }
-
-  void linkFileAndTable(TableId tableId, String file) {
-    // get the internal id for this table
-    int internalId = internalIds.get(tableId);
-
-    // Initialize a bitset for tables (internal IDs) that reference this file
-    Integer[] tables = tableFiles.get(file);
-    if (tables == null) {
-      tables = new Integer[internalIds.size()];
-      for (int i = 0; i < tables.length; i++)
-        tables[i] = 0;
-      tableFiles.put(file, tables);
-    }
-
-    // Update the bitset to track that this table has seen this file
-    tables[internalId] = 1;
-  }
-
-  void addFileSize(String file, long size) {
-    fileSizes.put(file, size);
-  }
-
-  Map<List<TableId>,Long> calculateUsage() {
-
-    // Bitset of tables that contain a file and total usage by all files that share that usage
-    Map<List<Integer>,Long> usage = new HashMap<>();
-
-    if (log.isTraceEnabled()) {
-      log.trace("fileSizes {}", fileSizes);
-    }
-    // For each file w/ referenced-table bitset
-    for (Entry<String,Integer[]> entry : tableFiles.entrySet()) {
-      if (log.isTraceEnabled()) {
-        log.trace("file {} table bitset {}", entry.getKey(), Arrays.toString(entry.getValue()));
-      }
-      List<Integer> key = Arrays.asList(entry.getValue());
-      Long size = fileSizes.get(entry.getKey());
-
-      Long tablesUsage = usage.get(key);
-      if (tablesUsage == null)
-        tablesUsage = 0L;
-
-      tablesUsage += size;
-
-      usage.put(key, tablesUsage);
-
-    }
-
-    Map<List<TableId>,Long> externalUsage = new HashMap<>();
-
-    for (Entry<List<Integer>,Long> entry : usage.entrySet()) {
-      List<TableId> externalKey = new ArrayList<>();
-      List<Integer> key = entry.getKey();
-      // table bitset
-      for (int i = 0; i < key.size(); i++)
-        if (key.get(i) != 0)
-          // Convert by internal id to the table id
-          externalKey.add(externalIds.get(i));
-
-      // list of table ids and size of files shared across the tables
-      externalUsage.put(externalKey, entry.getValue());
-    }
+/**
+ * Class that computes shared disk usage across tables by scanning HDFS files.
+ *
+ * @deprecated This class has been deprecated in favor of
+ *             {@link org.apache.accumulo.core.util.tables.MetadataTableDiskUsage}
+ */
+@Deprecated
+public class HdfsTableDiskUsage extends TableDiskUsage {
```

Comment:
I don't think we need subclasses for disk usage. That seems unnecessarily complex. I think when we get the list of files, we can just keep a pair instead of a set... the pair including the filename and the size. Then, get the size from that mapping, rather than lookup in HDFS.

I think a lot of the complexity of this PR is trying to preserve both old and new code... and that's not necessary.

Reply:
So the reason for the subclasses and the new API was @milleruntime mentioned not wanting to break the  thrift API which makes sense. Because new version is a client side scan only and doesn't need to make a server call I kept the thrift API alone and preserved the existing implementation. The other option would be I guess to move the metadata scan to the server and replace it so the thrift API call gets the new implantation but that seems unnecessary since you can do a client scan. Or I guess you could always have both, have it server wise in the API and also make it available in the client.

Also the new API was created since more information is returned with the new call to support the new output. However this is not a big deal if we don't want to change the output as you indicated in your other comment.

So mostly the issue is what to do about the existing server side API  that is public in thrift and what that should use and if we should keep the metadata scan only client  side.

Reply:
@ctubbsii  - Also in regards to your comment about complexity, the majority of the complexity was already there previously with the disk usage tracking since the purpose is to track shared disk usage. All I essentially did was take the existing TableDiskUsage implementation and break the common computation and data structures into a parent class and then have 2 subclasses to make it possible to switch out the source of the file sizes (HDFS or metadatable) while keeping the existing logic so the API doesn't change and the command still works the same way.

If the goal here is to make this a simple PR to simply switch out the source of the file size from being HDFS and to use metadata then would want to just revert back to what is currently there now in [TableDiskUsage](https://github.com/apache/accumulo/blob/87402dfcd8d828850ca2e4b01e8811cf506e7d4e/server/base/src/main/java/org/apache/accumulo/server/util/TableDiskUsage.java) and get rid of the subclasses and then just swap out the source of the file size from HDFS just to using the Metadata table and keep my code changes to do that. This way we know the logic still works and would work with the command, the API wouldn't change, and the other changes would be minimal.

But again, as I said in my previous comment, I can do that but there's still the issue of the client vs server side and thrift API call and whether or not we need to keep that and if we should be doing this client side vs server side scan as the old implementation runs on a server vs my new implementation that runs client side.